### ISSUE TITLE
unknown type name 'pthread_barrier_t' on Mac

### ISSUE THREAD

**Issue by astralbat** [Edrusb/libthreadar#2]
Hello,

I'm unable to compile this on Mac with the following make output:

make
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-recursive
Making all in src
depbase=`echo libthreadar.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
	/bin/sh ../libtool  --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I..     -g -O2 -std=c++11 -MT libthreadar.lo -MD -MP -MF $depbase.Tpo -c -o libthreadar.lo libthreadar.cpp &&\
	mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I.. -g -O2 -std=c++11 -MT libthreadar.lo -MD -MP -MF .deps/libthreadar.Tpo -c libthreadar.cpp  -D__DYNAMIC__  -fno-common -DPIC -o .libs/libthreadar.o
In file included from libthreadar.cpp:38:
In file included from ./libthreadar.hpp:52:
./barrier.hpp:92:2: error: unknown type name 'pthread_barrier_t'
        pthread_barrier_t bar;
        ^
1 error generated.
make[2]: *** [libthreadar.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

It seems pthread_barrier_t is not available on Mac?

**Comment by Edrusb** [comment]
you need to install Posix thread package which will bring you the pthread.h header file and libpthread library

**Comment by astralbat** [comment]
Thanks for getting back Denis.

I should already have this installed:

```
grep pthread config.log
configure:16927: checking for pthread_mutex_init in -lpthread
configure:16949: g++ -o conftest -g -O2 -std=c++11   conftest.cpp -lpthread   >&5
configure:17124: checking for pthread.h
configure:17369: g++ -o conftest -g -O2 -std=c++11   conftest.cpp -lpthread  >&5
ac_cv_header_pthread_h=yes
ac_cv_lib_pthread_pthread_mutex_init=yes
LIBS='-lpthread '
```
Nevertheless, I ran `xcode-select --install` again to make sure.

From looking around, it seems to me that barriers are an optional part of the pthread spec and Apple hasn't implemented them. See comment in http://mff.devnull.cz/pvu/src/pthreads/pthread-barrier.c.

Am I missing something or doesn't this compile on Macs? I was hoping to use remote functionality on dar.

Thanks.

**Comment by astralbat** [comment]
I think I may wind up using dar locally on an external HDD instead. It might work better for my use case if I'm doing frequent differentials.

**Comment by Edrusb** [comment]
ouch! Good finding!

I will have to find a workaround to barriers for MacOS, else libthreadar will never compile under Macs... 
Good to know you could find an alternative, but I have to make it working under MAC OS X anyway. 

Would you mind validating the possible adaptation I would make to workaround barrier? (I have no access to any MAC OS X platform)...

Cheers,
Denis

**Comment by astralbat** [comment]
Yes for sure I can help with that

**Comment by Edrusb** [comment]
Hi, could you test the release candidate 1.4.0.RC1 available either here on github using the branch "branch_1.4.x" or using prepackaged source code available at https://dar.edrusb.org/libthreadar/Release%20Candidates/ with associated gpg signature.

Thanks in advance for your help!

**Comment by astralbat** [comment]
Thanks. I have now got the following error:

> test_barrier.cpp:11:22: error: reference to 'thread' is ambiguous
> class myfile: public thread
>                      ^
> ./../../src/thread.hpp:97:11: note: candidate found by name lookup is 'libthreadar::thread'
>     class thread
>           ^
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__threading_support:404:24: note:
>       candidate found by name lookup is 'std::__1::thread'
> class _LIBCPP_TYPE_VIS thread;
>                        ^
> test_barrier.cpp:11:22: error: base class has incomplete type
> class myfile: public thread
>               ~~~~~~~^~~~~~
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__threading_support:404:24: note:
>       forward declaration of 'std::__1::thread'
> class _LIBCPP_TYPE_VIS thread;
>                        ^
> test_barrier.cpp:27:18: error: 'inherited_run' marked 'override' but does not override any
>       member functions
>     virtual void inherited_run() override
>                  ^
> test_barrier.cpp:22:2: error: no matching function for call to 'kill'
>         kill();
>         ^~~~
> /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/signal.h:80:5: note:
>       candidate function not viable: requires 2 arguments, but 0 were provided
> int     kill(pid_t, int) __DARWIN_ALIAS(kill);
>         ^
> test_barrier.cpp:23:2: error: use of undeclared identifier 'join'
>         join();
>         ^
> test_barrier.cpp:48:17: error: use of undeclared identifier 'make_unique'
>         file.push_back(make_unique<myfile>(&bar, i));
>                        ^
> test_barrier.cpp:51:9: error: no member named 'run' in 'myfile'
>         (*it)->run();
>         ~~~~~  ^
> test_barrier.cpp:54:9: error: no member named 'join' in 'myfile'
>         (*it)->join();
>         ~~~~~  ^
> 8 errors generated.
> make[3]: *** [test_barrier.o] Error 1
> make[2]: *** [all-recursive] Error 1
> make[1]: *** [all-recursive] Error 1
> make: *** [all] Error 2

**Comment by Edrusb** [comment]
OK, this is a conflict about the symbol "thread" between std and libthreadar namespaces used in this test_barrier.cpp module. I've fixed it (should be at least) in RC2 available beside RC1 as mentionned above

Thanks a lot for your help :)

**Comment by astralbat** [comment]
No problem. I'm just working from git if it makes things any easier.

Now. I get:

> test_barrier.cpp:47:17: error: use of undeclared identifier 'make_unique'
>         file.push_back(make_unique<myfile>(&bar, i));
>                        ^
> 1 error generated.
> make[3]: *** [test_barrier.o] Error 1
> make[2]: *** [all-recursive] Error 1
> make[1]: *** [all-recursive] Error 1
> make: *** [all] Error 2

**Comment by Edrusb** [comment]
OK, using git makes things easier for me too. Just pushed two new commits on branch_1.4.x : the first should fix the undeclared std::make_unique symbol (strangely under Linux the absence of <memory> include does not hurt...) Thank for you continued help!

**Comment by astralbat** [comment]
No problem. Strangely getting the same error a bit further down.

> test_barrier.cpp:72:17: error: use of undeclared identifier 'make_unique'
>         file.push_back(make_unique<myfile>(&bar, i));
>                        ^
> 1 error generated.
> make[3]: *** [test_barrier.o] Error 1
> make[2]: *** [all-recursive] Error 1
> make[1]: *** [all-recursive] Error 1
> make: *** [all] Error 2

**Comment by Edrusb** [comment]
OK found the probable reason: std::make_unique is a feature available since C++14, will avoid using it (especially in a testing program) to allow compilation to succeed for C++11 only compilers. Git is updated, all compiles under Linux, now I'm crossing fingers :)

**Comment by astralbat** [comment]
That's it, got it to compile, install and be detected by dar during configure.

Thanks very much.

**Comment by Edrusb** [comment]
Great news! Thanks for your help validating the compilation under Mac OS, Much appreciated!
I just have released version 1.4.0 
closing this issue, and thanks a gain for your help :)
